### ISSUE TITLE
Comparison with variances

### ISSUE THREAD

**Issue by jl-wynen** [scipp/scipp#1734]
Should comparisons include variances and if so, how? Currently, `is_equal` and `is_approx` ignore variances.

As far as (in)equality operators of `ValueAndVariance` are concerned, this makes sense in that the different operators are consistent, e.g. `a >= b && a < b` implies `a == b`. And it is not even possible to include the variances in less than and greater than without some serious statistical analysis.

But what do users expect the high level functions to do? For example, (by @OwenArnold)
```.py
b = a.copy()
a /= x
a *= x
is_approx(a, b)
```
The values should be approximately equal but the variances can be much larger in `b` than in `a`.

**Comment by jl-wynen** [comment]
What I would expect as a user is:
- `is_approx(a, b, tol)` checks if `a` and `b` equal to within floating point precision with given tolerance. And that means everything in the operands, values _and_ variances. There is a problem, though: An absolute tolerance might not be appropriate for both and we might have to resort to relative tolerances or two separate numbers, one for values and one for variances.
- A different function (maybe part of scipp, maybe not) `iswithin(a, b, sigma=5)` checks whether `a` and `b` are within the given number of standard deviations from each other.
- I don't see any practical situation in which we would want to just ignore the variances. But if one arises, one can just compare `.values` instead of the full variable. Or maybe there can be a separate function or a keyword argument of `is_approx`.

**Comment by jl-wynen** [comment]
Some thoughts on the example above. In this particular case, the only reason why the variances grow, is that we are ignoring correlations. If we included them, then the variances of a and b should be the same (within float precision).

In real life, if we have operations that cancel each other in the values, the precision of the result depends on the statistics. If the operands are uncorrelated, then a purely float based comparison is meaningless, we have to do a statistical check because the operations will only cancel to within statistical uncertainties. The only case where the float based `is_approx` makes sense is when the operands are fully correlated and the variances cancel or the operations do not actually combine values but only copy them around, e.g. binning.

**Comment by SimonHeybrock** [comment]
> Currently, `is_equal` and `is_approx` ignore variances.

`is_equal` does not? It maps to `operator==` in C++.

**Comment by SimonHeybrock** [comment]
One odd thing is the `is_approx` appears to ignore variances in the tolerance?

```python
import scipp as sc
a = sc.scalar(value=1.0, variance=1.0)
b = sc.scalar(value=1.0, variance=2.0)
sc.is_approx(a, b, sc.scalar(value=0.1, variance=0.1))
```
givens `True`.

**Comment by SimonHeybrock** [comment]
It should also be noted that `is_approx` is mostly meant for unit tests. Maybe it should be removed from the "public" (documented?) API? Or moved into a submodule?

**Comment by jl-wynen** [comment]
> is_equal does not? It maps to operator== in C++.

https://github.com/scipp/scipp/blob/fe1f727eefe89f293cb7e9cb4112cbcfd5eb5ad8/core/include/scipp/core/value_and_variance.h#L190

> It should also be noted that is_approx is mostly meant for unit tests. Maybe it should be removed from the "public" (documented?) API? Or moved into a submodule?

Could still be useful for users to test their stuff.

**Comment by OwenArnold** [comment]
> It should also be noted that `is_approx` is mostly meant for unit tests. Maybe it should be removed from the "public" (documented?) API? Or moved into a submodule?

Mainly yes, however I can imagine a user moving from say mantid (or something else) to scipp and maybe wanting to verify closeness natively.

**Comment by SimonHeybrock** [comment]
> > It should also be noted that `is_approx` is mostly meant for unit tests. Maybe it should be removed from the "public" (documented?) API? Or moved into a submodule?
> 
> Mainly yes, however I can imagine a user moving from say mantid (or something else) to scipp and maybe wanting to verify closeness natively.

I think we should probably still consider making this part of a "detail" API. `sc.isclose` (meta data behavior similar to other binary operations), but maybe it should be `sc.detail.is_approx` for the fuzzy testing? Or something like `sc.testing.is_near`? My point is that we should distinguish between things a user would want to use in production code (things that, e.g., have a numpy equivalent), vs. more specialized functions?

**Comment by SimonHeybrock** [comment]
> > is_equal does not? It maps to operator== in C++.
> 
> https://github.com/scipp/scipp/blob/fe1f727eefe89f293cb7e9cb4112cbcfd5eb5ad8/core/include/scipp/core/value_and_variance.h#L190

That is NOT being used for `Variable::operator==`, instead:

https://github.com/scipp/scipp/blob/fe1f727eefe89f293cb7e9cb4112cbcfd5eb5ad8/variable/include/scipp/variable/data_model.h#L139-L158

**Comment by jl-wynen** [comment]
The question here is not whether we should do fuzzy matching on metadata. We want to know whether to include variances. By `is_approx` I meant the function that we have now which is supposed to turn into `isclose`.

> That is NOT being used for Variable::operator==, instead:

Ok, but `is_approx` does ignore variances at the moment because it calls `operator<=` on the elements: 
https://github.com/scipp/scipp/blob/fe1f727eefe89f293cb7e9cb4112cbcfd5eb5ad8/core/include/scipp/core/element/comparison.h#L31

**Comment by SimonHeybrock** [comment]
Ok, for the rest of the dicussion, to avoid further confusion:
- Let us refer to what we want to be the equivalent of `np.isclose` as `isclose` (whether it would be the current `is_approx` or not).
- Fuzzy comparse that allows fuzzy *coord* matching as `near` (mnemonic: `EXPECT_NEAR` from `gtest`).

**Comment by SimonHeybrock** [comment]
Regarding `isclose`:
- I see this as being in the same group of operations as `__equal__`, `__less__`, ..., i.e., it should (exectly) compare coords on `OR` masks.
- The `data` of the ouput has `dtype=bool`, so if the inputs contain variances we still need to return elements (true or false) without variances.
- As you say above, without knowing correlations, it does not make sense to consider variances when comparing values.
- It feels like the only options we have left therefore are:
  1. Ignore variances (same as `is_approx` now).
  2. Throw an exception if there are variances (not useful?). User could run `sc.isclose(sc.values(data1), sc.values(data2))`.
  3. Return true if `value_delta < tol AND variance_delta < tol` (or `tol^2` in the latter?)

**Comment by OwenArnold** [comment]
`isclose` implemented, which will compare variances. Implemented as part of #1743
